L2SIVacuum/POUs/Functions/Valves/FB_VRC_NC_ClsLS.TcPOU

```diff
@@ -0,0 +1,264 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.18">
+  <POU Name="FB_VRC_NC_ClsLS" Id="{c8e0c1f7-5d63-4e08-9f94-8c1d7ca472c0}" SpecialFunc="None">
+    <Declaration><![CDATA[(* This Function Block Implements Basic Functionality for ChemRIX NC valve, which has one NC LS mounting at Open position. 
+Valve considers CLOSED when the LS is closed. 
+This function block is interloked by an input (i_xExtILK_OK). This is so developers can interface with custom
+interlocking logic outside this function block.*)
+(* Note Interlock Logic is External *)
+{attribute 'no_check'}
+FUNCTION_BLOCK FB_VRC_NC_ClsLS EXTENDS FB_Valve
+VAR_IN_OUT
+	
+END_VAR
+VAR_INPUT
+	i_xExtILK_OK : BOOL; //Connect to Interlock logic condition(e.g F_TURBO_VRC_ILK Function), otherwise, Set to True if the valve is not interlocked
+	i_xOverrideMode : BOOL; (*To be linked to global override bit. This Overrides Vacuum logic only, EPS, MPS and PMPS are still enforces*)
+END_VAR
+VAR_OUTPUT
+	{attribute 'pytmc' := '
+    pv:
+	'}
+	iq_stValve : ST_VRC;
+END_VAR
+VAR
+	{attribute 'instance-path'}
+	{attribute 'noinit'}
+	sPath: STRING;
+	xFirstPass	:	BOOL;
+	fbFSInit		:	R_TRIG;
+	tonOvrd	:	TON;
+	tonDelOK : TON;
+	rtOK	:	R_TRIG;
+	tOvrd	:	TIME := T#10s;
+	(* Timeouts*)
+	tTimeOutDuration: TIME:= T#30S;
+	tOPNtimeout: TON;
+	tCLStimeout:TON;
+	
+	
+	(*IO*)
+//	i_xOpnLS	AT%I*: BOOL;
+	i_xClsLS	AT%I*: BOOL;
+	q_xOPN_DO	AT%Q*: BOOL;
+
+END_VAR]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[
+(* On first PLC pass, put valve into vented state, which implies a closed valve *)
+fbFSInit( CLK := TRUE, Q => xFirstPass);
+IF xFirstPass THEN
+	iq_stValve.eVGC_State := Vented;
+	iq_stValve.pv_xOPN_SW := FALSE;
+	ePrevState := INVALID; 
+END_IF
+
+
+
+
+///Check valve position
+(*IF iq_stValve.i_xClsLS AND  iq_stValve.i_xOpnLS THEN
+	iq_stValve.eState:=INVALID;
+ELSIF NOT iq_stValve.i_xClsLS AND iq_stValve.i_xOpnLS AND iq_stValve.q_xOPN_DO THEN 
+	iq_stValve.eState:=OPEN;
+ELSIF NOT iq_stValve.i_xClsLS AND iq_stValve.i_xOpnLS AND NOT iq_stValve.q_xOPN_DO THEN 
+	iq_stValve.eState:=OPEN_F;
+ELSIF iq_stValve.i_xClsLS AND NOT iq_stValve.i_xOpnLS AND NOT iq_stValve.q_xOPN_DO THEN 
+	iq_stValve.eState:=CLOSED;
+ELSIF NOT iq_stValve.i_xClsLS AND NOT iq_stValve.i_xOpnLS THEN 
+	iq_stValve.eState:=MOVING;
+ELSE
+	iq_stValve.eState:=INVALID;
+END_IF
+*)
+
+IF NOT iq_stValve.i_xClsLS AND iq_stValve.q_xOPN_DO THEN 
+	iq_stValve.eState:=OPEN;
+ELSIF NOT iq_stValve.i_xClsLS AND NOT iq_stValve.q_xOPN_DO THEN 
+	iq_stValve.eState:=OPEN_F;
```

Comment:
Not Sure we can have OPEN_F state. This might be an invalid state. We are not reading the CLS_LS but we are not opening it either. The cable might be unplugged?!

Reply:
Good point.  The valve cannot force open.  It should be invalid state.